Modules/_json.c

```diff
@@ -303,6 +303,89 @@ escape_unicode(PyObject *pystr)
     return rval;
 }
 
+// Take a PyUnicode pystr and write an ASCII-only escaped string to writer.
+static int
+write_escaped_ascii(PyUnicodeWriter *writer, PyObject *pystr)
+{
+    Py_ssize_t i;
+    Py_ssize_t input_chars;
+    Py_ssize_t chars;
+    Py_ssize_t copy_len = 0;
+    const void *input;
+    int kind;
+    int ret;
+    unsigned char buf[12];
+
+    input_chars = PyUnicode_GET_LENGTH(pystr);
+    input = PyUnicode_DATA(pystr);
+    kind = PyUnicode_KIND(pystr);
+
+    ret = PyUnicodeWriter_WriteChar(writer, '"');
+    if (ret) return ret;
+
+    for (i = 0; i < input_chars; i++) {
+        Py_UCS4 c = PyUnicode_READ(kind, input, i);
+        if (S_CHAR(c)) {
+            copy_len++;
+        }
+        else {
+            ret = PyUnicodeWriter_WriteSubstring(writer, pystr, i-copy_len, i);
+            if (ret) return ret;
+            copy_len = 0;
+
+            chars = ascii_escape_unichar(c, buf, 0);
+            ret = PyUnicodeWriter_WriteUTF8(writer, (const char*)buf, chars);
+            if (ret) return ret;
+        }
+    }
+
+    ret = PyUnicodeWriter_WriteSubstring(writer, pystr, i-copy_len, i);
+    if (ret) return ret;
+
+    return PyUnicodeWriter_WriteChar(writer, '"');
+}
+
+// Take a PyUnicode pystr and write an escaped string to writer.
+static int
+write_escaped_unicode(PyUnicodeWriter *writer, PyObject *pystr)
+{
+    Py_ssize_t i;
+    Py_ssize_t input_chars;
+    Py_ssize_t chars;
+    Py_ssize_t copy_len = 0;
+    const void *input;
+    int kind;
+    int ret;
+    unsigned char buf[12];
+
+    input_chars = PyUnicode_GET_LENGTH(pystr);
+    input = PyUnicode_DATA(pystr);
+    kind = PyUnicode_KIND(pystr);
+
+    ret = PyUnicodeWriter_WriteChar(writer, '"');
+    if (ret) return ret;
+
+    for (i = 0; i < input_chars; i++) {
+        Py_UCS4 c = PyUnicode_READ(kind, input, i);
+        if (c <= 0x1f || c == '\\' || c == '"') {
```

Comment:
I believe we can make this a single function with this extra condition:

```suggestion
        if (c <= 0x1f || c == '\\' || c == '"' || c >= 0x7f && ensure_ascii) {
```

Reply:
This function is PyUnicodeWriter version of escape_unicode.
In other words, this function implies ensure_ascii is used when Encoder is created.

Reply:
Yes, but this way we could make `write_escaped_ascii()` and `write_escaped_unicode()` a single function.